library/core/src/primitive_docs.rs

```diff
@@ -1527,6 +1527,76 @@ mod prim_ref {}
+///
+/// For two signatures to be considered *ABI-compatible*, they must use a compatible call ABI (as
+/// declared via `extern "ABI"`), must take the same number of arguments, and the individual
+/// argument types and the return types must be ABI-compatible.
+///
+/// The call ABIs are guaranteed to be compatible if they are the same, or if the caller ABI is
+/// `$X-unwind` and the callee ABI is `$X`.
+///
+/// The following types are guaranteed to be ABI-compatible:
+///
+/// - `*const T`, `*mut T`, `&T`, `&mut T`, `Box<T>` (specifically, only `Box<T, Global>`),
+///   `NonNull<T>` are all ABI-compatible with each other for all `T`. Two of these pointer types
+///   with different `T` are ABI-compatible if they have the same metadata type (`<T as
+///   Pointee>::Metadata`).
+/// - `usize` is ABI-compatible with the `uN` integer type of the same size, and likewise `isize` is
+///   ABI-compatible with the `iN` integer type of the same size.
+/// - Any two `fn`/`extern "ABI" fn` types are ABI-compatible with each other, independent of their
+///   signature.
+/// - Any two types with size 0 and alignment 1 are ABI-compatible.
+/// - A `repr(transparent)` type `T` is ABI-compatible with its unique non-trivial field, i.e., the
+///   unique field that doesn't have size 0 and alignment 1 (if there is such a field).
+/// - `i32` is ABI-compatible with `NonZeroI32`, and similar for all other integer types with their
+///   matching `NonZero*` type.
+/// - If `T` is guaranteed to be subject to the [null pointer
+///   optimization](option/index.html#representation), then `T` and `Option<T>` are ABI-compatible.
+///
+/// Furthermore, ABI compatibility satisfies the following general properties:
+///
+/// - Every type is ABI-compatible with itself.
+/// - If `T1` and `T2` are ABI-compatible, then two `repr(C)` types that only differ because one
+///   field type was changed from `T1` to `T2` are ABI-compatible.
+/// - If `T1` and `T2` are ABI-compatible and `T2` and `T3` are ABI-compatible, then so are `T1` and
+///   `T3` (i.e., ABI-compatibility is transitive).
+/// - If `T1` and `T2` are ABI-compatible, then so are `T2` and `T1` (i.e., ABI-compatibility is
+///   symmetric).
+///
+/// More signatures can be ABI-compatible on specific targets, but that should not be relied upon
+/// since it is not portable and not a stable guarantee.
+///
+/// Noteworthy cases of types *not* being ABI-compatible in general are `bool` vs `u8`, and `i32` vs
+/// `u32`: on some targets, the calling conventions for these types differ in terms of what they
+/// guarantee for the remaining bits in the register that are not used by the value. `i32` vs `f32`
+/// are not compatible either, as has already been mentioned above. `struct Foo(u32)` and `u32`
+/// are not compatible (without `repr(transparent)`) since structs are aggregate types and often
+/// passed in a different way than primitives like `i32`.
```

Comment:
```suggestion
/// Noteworthy cases of types *not* being ABI-compatible in general are:
/// * `bool` vs `u8`, and `i32` vs `u32`: on some targets, the calling conventions for these types differ in terms of what they
///   guarantee for the remaining bits in the register that are not used by the value.
/// * `i32` vs `f32` are not compatible either, as has already been mentioned above.
/// * `struct Foo(u32)` and `u32` are not compatible (without `repr(transparent)`) since structs are aggregate types and often
///   passed in a different way than primitives like `i32`.
```
edit to taste, but a list seems more organized here than just prose